File: example/ios/RNImageStitcherExample/FrameBrightnessPlugin.swift
// SPDX-License-Identifier: Apache-2.0
//
// FrameBrightnessPlugin — example-app demonstration of the v0.19.0
// native AR plugin framework (`RNISARFramePlugin`).
//
// The SDK ships ONLY the generic plugin framework — no concrete plugins.
// This sample proves the framework end-to-end: it computes the mean luma
// (0..1) of each ARFrame's `capturedImage` and returns it via the SYNC
// channel, so the JS `onArFrame` callback sees
// `meta.plugins.frameBrightness.brightness`.
//
// Registered once at startup in `AppDelegate` via
// `RNISARPluginRegistry.shared.register(...)`.  A real host plugin (e.g.
// RetaiLens's OCR) would offload heavy work to its own queue and use the
// ASYNC channel (`RNISARPluginRegistry.shared.emit(...)`) instead.

import Foundation
import CoreVideo
import RNImageStitcher

final class FrameBrightnessPlugin: NSObject, RNISARFramePlugin {

    func name() -> String { "frameBrightness" }

    /// Compute mean luma of the frame on the AR thread and return it
    /// synchronously.  This is intentionally cheap (sub-millisecond on a
    /// downsampled luma plane) so it's safe to run inline per frame.
    ///
    /// ARFrame.capturedImage is a bi-planar YUV
    /// (`kCVPixelFormatType_420YpCbCr8BiPlanarFullRange`) buffer; plane 0
    /// is the full-resolution Y (luma) plane.  We average a sparse,
    /// strided sample of plane 0 — averaging every pixel at 60Hz would be
    /// wasteful and unnecessary for a brightness readout.
    func process(_ context: RNISARFrameContext) -> [String: Any]? {
        let pb = context.pixelBuffer

        // Only the luma plane of a bi-planar YUV buffer is needed.  Bail
        // gracefully (nil → no result this frame) for any unexpected
        // format rather than misreading bytes.
        let format = CVPixelBufferGetPixelFormatType(pb)
        guard format == kCVPixelFormatType_420YpCbCr8BiPlanarFullRange ||
              format == kCVPixelFormatType_420YpCbCr8BiPlanarVideoRange else {
            return nil
        }

        // pixelBuffer is valid only during process(_:) — read it here,
        // synchronously, and copy out only the scalar result.
        CVPixelBufferLockBaseAddress(pb, .readOnly)
        defer { CVPixelBufferUnlockBaseAddress(pb, .readOnly) }

        guard let base = CVPixelBufferGetBaseAddressOfPlane(pb, 0) else {
            return nil
        }
        let width = CVPixelBufferGetWidthOfPlane(pb, 0)
        let height = CVPixelBufferGetHeightOfPlane(pb, 0)
        let rowBytes = CVPixelBufferGetBytesPerRowOfPlane(pb, 0)
        guard width > 0, height > 0, rowBytes > 0 else { return nil }

        let ptr = base.assumingMemoryBound(to: UInt8.self)

        // Sparse sampling: ~32 columns × ~32 rows across the frame keeps
        // this O(1024) regardless of capture resolution.
        let xStep = max(1, width / 32)
        let yStep = max(1, height / 32)
        var sum: UInt64 = 0
        var count: UInt64 = 0
        var y = 0
        while y < height {
            let row = ptr + y * rowBytes
            var x = 0
            while x < width {
                sum += UInt64(row[x])
                count += 1
                x += xStep
            }
            y += yStep
        }
        guard count > 0 else { return nil }

        // Mean luma normalised to 0..1 (8-bit luma → /255).
        let brightness = Double(sum) / Double(count) / 255.0
        return ["brightness": brightness]
    }
}

File: example/android/app/src/main/java/com/rnimagestitcherexample/FrameBrightnessPlugin.kt
package com.rnimagestitcherexample

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.WritableMap
import io.imagestitcher.rn.ARFrameContext
import io.imagestitcher.rn.ARFramePlugin

/**
 * 0.19.0 — TINY sample [ARFramePlugin] proving the AR plugin framework
 * end-to-end on Android (Swift twin: example/ios `FrameBrightnessPlugin`).
 *
 * The SDK ships ONLY the generic framework (no OCR / no concrete plugins) —
 * this lives in the EXAMPLE app to demonstrate the host-side contract:
 * implement [ARFramePlugin], register it with `RNSARPluginRegistry` at
 * startup, read the result in JS.
 *
 * Computes the mean luma (Y plane) of each ARCore frame and returns a SYNC
 * result `{ "brightness": <0..1> }`.  The SDK folds this into the throttled
 * `onArFrame` event under `meta.plugins.frameBrightness`, which the example
 * `App.tsx` surfaces in its on-screen AR overlay.
 *
 * ## Why sync (not async)
 *
 * Mean luma is microseconds of arithmetic — cheap enough to run inline on
 * the AR thread and return synchronously.  Heavier plugins (OCR, ML) would
 * instead return `null`, COPY the bytes they need (`ctx.nv21.copyOf()`),
 * offload to their own queue, and later call
 * `RNSARPluginRegistry.emit("...", result)` to deliver over the async
 * `RNImageStitcherARPluginResult` event.
 *
 * ## Subsampling
 *
 * We sample every Nth pixel ([STRIDE]) of the Y plane rather than summing
 * all ~2M luma bytes per frame — mean brightness is stable under heavy
 * subsampling and this keeps the per-frame cost trivial.
 */
class FrameBrightnessPlugin : ARFramePlugin {

    override fun name(): String = PLUGIN_NAME

    override fun process(context: ARFrameContext): WritableMap? {
        val w = context.width
        val h = context.height
        val ySize = w * h
        if (ySize <= 0 || context.nv21.size < ySize) return null

        // Mean luma over a strided sample of the Y plane.  Y bytes are
        // unsigned 0..255 stored in signed Kotlin Bytes → mask with 0xFF.
        // Read directly from the NV21 array (Y plane is the first ySize
        // bytes) — no copy needed since we consume it synchronously here.
        val nv21 = context.nv21
        var sum = 0L
        var count = 0
        var i = 0
        while (i < ySize) {
            sum += (nv21[i].toInt() and 0xFF)
            count++
            i += STRIDE
        }
        if (count == 0) return null

        val brightness = (sum.toDouble() / count) / 255.0  // 0..1

        return Arguments.createMap().apply {
            putDouble("brightness", brightness)
        }
    }

    companion object {
        /// Plugin name — the key under `meta.plugins` on the JS side.
        const val PLUGIN_NAME = "frameBrightness"

        /// Sample every Nth Y-plane byte (mean luma is stable under heavy
        /// subsampling; keeps per-frame cost trivial).
        private const val STRIDE = 64
    }
}
